File: College/Second Year/DSA/Ex2_.InfixToPostfix.py
print("╔═════════════╗\n║   Tanvik    ║\n║ URK23CS1261 ║\n╚═════════════╝")
class Main:
    def __init__(self):
        self.expression = ""
        self.operators = ['+', "-", "/","*", "^"]
        self.stack = []
        self.output = []

    def input_expression(self):
        self.expression = (input("Enter the Expression: ")).replace(" ", "")
        self.stack = []
        self.output = []

    def InfixToPostfix(self):
        def precedence(element: str):
            if element == '^':
                return 3
            elif element == '*' or element == '/':
                return 2
            elif element == '+' or element == '-':
                return 1
            else:
                return 0
        for index in range(len(self.expression)):
            char = self.expression[index]
            if char.isalnum():
                self.output.append(char)
            elif char == "(":
                self.stack.append(char)
            elif char in self.operators:
                if self.stack and self.stack[-1] != '(':
                    if precedence(self.stack[-1]) > precedence(char):
                        self.output.append(self.stack[-1])
                        self.stack[-1] = char
                    elif precedence(self.stack[-1]) < precedence(char):
                        self.stack.append(char)
                    elif precedence(self.stack[-1]) == precedence(char):
                        self.output.append(self.stack[-1])
                        self.stack[-1] = char
                else:
                    self.stack.append(char)
            elif char == ')':
                while self.stack and self.stack[-1] != '(':
                    self.output.append(self.stack.pop())
                self.stack.pop()
        while self.stack:
            self.output.append(self.stack[-1])    
            self.stack.pop()
        print("Postfix Expression: ",*self.output, sep='')
    def EvaluatePostfix(self):
        for char in self.expression:
            if char.isdigit():
                self.stack.append(int(char))
            elif char in self.operators:
                if len(self.stack) < 2:
                    print("Error: Invalid expression")
                    return
                op1 = self.stack.pop()
                op2 = self.stack.pop()
                if char == '+':
                    self.stack.append(op2 + op1)
                elif char == '-':
                    self.stack.append(op2 - op1)
                elif char == '*':
                    self.stack.append(op2 * op1)
                elif char == '/':
                    self.stack.append(op2 / op1)
                elif char == '^':
                    self.stack.append(op2 ** op1)
        print("Postfix Expression Evaluation:", self.stack[0])
instance = Main()
# instance.InfixToPostfix()
instance.input_expression()
#231*+9-
#100 200 + 2 / 5 * 7 +
instance.EvaluatePostfix()
#A+ (B*C-(D/E^F)*G)*H == ABC*DEF^/G*-H*+
#A+B-(C*D-E)+F == AB+CD*E--F+

File: College/Second Year/DSA/Ex2_InfixToPostfix.java
import java.util.Scanner;
import java.util.Stack;

public class Ex2_InfixToPostfix {
    private String expression = "";
    private final String operators = "+-/*^";
    private final Stack<Character> operatorStack;
    
    public Ex2_InfixToPostfix(String expression){
        this.expression = expression.replaceAll("\\s+", "");
        this.operatorStack = new Stack<>();
    }
    private int precedence(char operator) {
        return switch (operator) {
            case '^' -> 3;
            case '*', '/' -> 2;
            case '+', '-' -> 1;
            default -> 0;
        };
    }
    public void infixToPostfix() {
        StringBuilder output = new StringBuilder();
        for (int i = 0; i < expression.length(); i++) {
            char current = expression.charAt(i);
            if (Character.isLetterOrDigit(current)) {
                output.append(current);
            } else if (current == '(') {
                operatorStack.push(current);
            } else if (operators.indexOf(current) != -1) {
                while (!operatorStack.isEmpty() && operatorStack.peek() != '('
                && ((precedence(operatorStack.peek()) >= precedence(current)) && current != '^')) {
                    output.append(operatorStack.pop());
                }
                operatorStack.push(current);
            } else if (current == ')') {
                while (!operatorStack.isEmpty() && operatorStack.peek() != '(') {
                    output.append(operatorStack.pop());
                }
                if (!operatorStack.isEmpty() && operatorStack.peek() == '(') {
                    operatorStack.pop();
                }
            }
        }

        while (!operatorStack.isEmpty()) {
            output.append(operatorStack.pop());
        }
        System.out.println("Postfix Expression: " + output.toString());
    }
    public void evaluatePostfix() {
        Stack<Integer> evaluationStack = new Stack<>();
        for (int i = 0; i < expression.length(); i++) {
            char current = expression.charAt(i);
            if (Character.isDigit(current)) {
                evaluationStack.push(Character.getNumericValue(current));
            } else if (operators.indexOf(current) != -1) {
                if (evaluationStack.size() < 2) {
                    System.out.println("Error: Invalid expression");
                    return;
                }
                int op1 = evaluationStack.pop();
                int op2 = evaluationStack.pop();
                switch (current) {
                    case '+' -> evaluationStack.push(op2 + op1);
                    case '-' -> evaluationStack.push(op2 - op1);
                    case '*' -> evaluationStack.push(op2 * op1);
                    case '/' -> evaluationStack.push(op2 / op1);
                    case '^' -> evaluationStack.push((int) Math.pow(op2, op1));
                }
            }
        }
        if (!evaluationStack.isEmpty()) {
            System.out.println("Postfix Expression Evaluation: " + evaluationStack.peek());
        }
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("\nMenu:\n1. Infix to Postfix\n2. Evaluate Postfix Expression\n3. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            scanner.nextLine(); 
            switch (choice) {
                case 1 -> {
                    System.out.print("Enter the Expression: ");
                    String expression = scanner.nextLine();
                    Ex2_InfixToPostfix instance = new Ex2_InfixToPostfix(expression);
                    instance.infixToPostfix();
                }
                case 2 -> {
                    System.out.print("Enter the Expression: ");
                    String expression = scanner.nextLine();
                    Ex2_InfixToPostfix instance = new Ex2_InfixToPostfix(expression);
                    instance.evaluatePostfix();
                }
                case 3 -> {
                    System.out.println("Exiting...");
                    System.exit(0);
                    return;
                }
                default -> System.out.println("Invalid choice. Please try again.");
            }
        }
    } 
}
